src/client/auth/aws.rs

```diff
@@ -0,0 +1,402 @@
+                "MONGODB-AWS",
+                "cannot specify only one of access key and secret key; either both or neither \
+                 must be provided",
+            ));
+        }
+
+        if session_token.is_some() {
+            return Err(Error::authentication_error(
+                "MONGODB-AWS",
+                "cannot specify session token without both access key and secret key",
+            ));
+        }
+
+        if let Ok(relative_uri) = std::env::var("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI") {
+            Self::get_from_ecs(relative_uri, http_client).await
+        } else {
+            Self::get_from_ec2(http_client).await
+        }
+    }
+
+    async fn get_from_ecs(relative_uri: String, http_client: &HttpClient) -> Result<Self> {
+        // Use the local IP address that AWS uses for ECS agents.
+        let uri = format!("http://169.254.170.2/{}", relative_uri);
+
+        http_client.get_and_deserialize_json(&uri, None).await
+    }
+
+    async fn get_from_ec2(http_client: &HttpClient) -> Result<Self> {
+        let temporary_token = http_client
+            .put_and_read_string(
+                "http://169.254.169.254/latest/api/token",
+                &[("X-aws-ec2-metadata-token-ttl-seconds", "30")],
+            )
+            .await?;
+
+        let role_name_uri = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+
+        let role_name = http_client
+            .get_and_read_string(
+                role_name_uri,
+                &[("X-aws-ec2-metadata-token", &temporary_token[..])],
+            )
+            .await?;
+
+        let credential_uri = format!("{}/{}", role_name_uri, role_name);
+
+        http_client
+            .get_and_deserialize_json(
+                &credential_uri,
+                &[("X-aws-ec2-metadata-token", &temporary_token[..])],
+            )
+            .await
+    }
+
+    fn compute_authorization_header(
+        &self,
+        date: DateTime<Utc>,
+        host: &str,
+        server_nonce: &[u8],
+    ) -> Result<String> {
+        let date_str = date.format("%Y%m%dT%H%M%SZ").to_string();
+
+        // We need to include the security token header if the user provided a token. If not, we
+        // just use the empty string.
+        let token = self
+            .session_token
+            .as_ref()
+            .map(|s| format!("x-amz-security-token:{}\n", s))
+            .unwrap_or_default();
+
+        // Similarly, we need to put "x-amz-security-token" into the list of signed headers if the
+        // user provided a token. If not, we just use the empty string.
+        let token_signed_header = if self.session_token.is_some() {
+            "x-amz-security-token;"
+        } else {
+            ""
+        };
+
+        // Generate the list of signed headers (either with or without the security token header).
+        #[rustfmt::skip]
+        let signed_headers = format!(
+            "\
+              content-type;\
```

Comment:
These headers need to be in alphabetical order
https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html